scanner/updater/jsonblob/jsonblob.go

```diff
@@ -0,0 +1,152 @@
+// Package jsonblob is a fork of ClairCore's jsonblob with minimal changes to
+// support individual record iteration.
+//
+// TODO(ROX-24333): Remove this package once ClairCore supports it, or become
+//                  obsolete by other formats.
+
+package jsonblob
+
+import (
+	"bufio"
+	"encoding/json"
+	"io"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/libvuln/driver"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+type iter2[X, Y any] func(yield func(X, Y) bool)
+
+// RecordIter iterates over records of an update operation.
+type RecordIter iter2[*claircore.Vulnerability, *driver.EnrichmentRecord]
+
+// OperationIter iterates over operations, offering a nested iterator for records.
+type OperationIter iter2[*driver.UpdateOperation, RecordIter]
+
+// Iterate iterates over each record serialized in the [io.Reader] grouping by
+// update operations. It returns an OperationIter, which is an iterator over each
+// update operation with a nested iterator for the associated vulnerability
+// entries, and an error function, to check for iteration errors.
+func Iterate(r io.Reader) (OperationIter, func() error) {
+	var err error
+	var de diskEntry
+
+	d := json.NewDecoder(r)
+	err = d.Decode(&de)
+
+	it := func(yield func(*driver.UpdateOperation, RecordIter) bool) {
+		for err == nil {
+			op := &driver.UpdateOperation{
+				Ref:         de.Ref,
+				Updater:     de.Updater,
+				Fingerprint: de.Fingerprint,
+				Date:        de.Date,
+				Kind:        de.Kind,
+			}
+			it := func(yield func(*claircore.Vulnerability, *driver.EnrichmentRecord) bool) {
+				var vuln *claircore.Vulnerability
+				var en *driver.EnrichmentRecord
+				for err == nil && op.Ref == de.Ref {
+					vuln, en, err = de.Unmarshal()
+					if err != nil || !yield(vuln, en) {
```

Comment:
From what I can tell, this all works under the assumption that each update operation only references one of vulns and enrichments, right?

Reply:
Yes, as dictated by its type.